### PR TITLE
Fixes #36104 - Distinguish host events

### PR DESCRIPTION
**Pull request by ofedoren**
Testable with `foreman_webhooks`. Adds new events to better distinguish between host events, whether it was triggered by user or via some automation.

@adamruzicka, what do you think? This is the simplest solution I could find and it seems to solve the issue...

### CODE REVIEW

**Comment by ofedoren** [comment]
Added a second commit, which removes new `host_*[_automatically|_manually]` events. Host created/destroyed can be left as general events, since they are pretty rare in comparison to host updated event we want to resolve problems with (due to often auto updates via e.g. reports). This commit ensures `host_updated` event is only triggered on actual user input against the host object. Auto updates (such as reports or changes through console) won't or shouldn't trigger this event. This comes from offline discussion with @ares, but any other input is surely welcome.

**Comment by adamruzicka** [comment]
I like this, it is nice and clean and seems to work well. However, shouldn't there be an event that would fire once uploaded facts get processed?

**Comment by ofedoren** [comment]
@adamruzicka, that event can cause spammed `host_updated` events. If we drop the second commit then it will fire `host_updated_automatically` event, which nobody used before, so it's pretty safe. But in the same time, `host_updated` will be still general, so users would need to manually re-subscribe webhooks to a different event.

I agree that with the last change here, users can no longer receive events if a host was updated by facts upload, but at the same time, if there will be a request, we could add either a custom `facts_upload` event, or `host_updated_facts` or even again `host_updated_automatically`.

Other than that, I'd personally prefer the initial fix (without second commit), but I can be missing something, so I'd like to see what direction here we would all like to take, @adamruzicka, @ares.

**Comment by adamruzicka** [comment]
> we could add either a custom facts_upload event, or host_updated_facts

I had something like that in my mind. Spamming events is definitely wrong, but not firing anything when facts get uploaded doesn't feel right either

**Comment by ofedoren** [comment]
@adamruzicka, I've added `host_facts_updated` event, but I guess it will work only on `Host::Managed`.

It seems we'd need to check how webhooks work for unmanaged and discovered hosts (are there any other?).

File: app/services/host_fact_importer.rb [hunk 1]
```diff
@@ -31,7 +31,9 @@ def import_facts(facts, source_proxy = nil)
       host.save(:validate => false)
     end
 
-    parse_facts(facts, type, source_proxy)
+    state = parse_facts(facts, type, source_proxy)
+    host.try(:trigger_hook, :host_facts_updated)
```
Comment (adamruzicka) [inline app/services/host_fact_importer.rb]: Can `Host::Managed#trigger_hook` raise an exception? If so, should we wrap this in begin+rescue to keep the main flow working even if even triggering does not?

**Review by adamruzicka — COMMENTED** [review]
Works well, one inline comment

**Comment by ofedoren** [comment]
Thanks, @adamruzicka, updated. AFAIK it could throw `NoMethodError`, but apart from that I'm not sure it can raise something else, but just to be sure there is a catch for any exception.

**Review by adamruzicka — APPROVED** [review]
Looks good to me, let's get this in

**Comment by adamruzicka** [comment]
Katello failures seem unrelated

**Comment by ofedoren** [comment]
Yeap: https://github.com/Katello/katello/pull/10701

[test katello]

**Comment by adamruzicka** [comment]
Could you please squash the commits before we merge?

**Comment by ofedoren** [comment]
Sure, done.

**Comment by adamruzicka** [comment]
Test failures are unrelated and are being resolved in another PR. Let's get this in

**Comment by adamruzicka** [comment]
Thank you @ofedoren !

**Pull request merged**